add command for usages out of the editor

YueHui opened the pull request.

robot-deployer committed:
add commandFile

robot-deployer committed:
change to command

YueHui committed:
Update README.md

YueHui committed:
add chinese translate

antoniolucasnobar committed:
[#1] Fix Typo

Commit:
[#1] remove duplicated method and format some code

Commit:
[#1] move removeSearchMarkers methods to utils:

- also removed unused variables and imports

Commit:
[#1] improve peerDependencies compatibility

Commit:
[#1] code formatting

Commit:
[#1] remove duplicated method

antoniolucasnobar commented:
Hello @YueHui !

First of all, thanks for your contribution. Create a command is a must for a ckeditor plugin!

I made some minor adjustments, removed some duplicated code and some code formatting. If you please could see if it is everything working fine in your environment, it would be great.

I let one or two notes in the code, if you please could look into, I appreciate.

antoniolucasnobar commented on src/findReplace.js:
```diff
@@ -219,147 +220,39 @@ export default class FindReplace extends Plugin {
+			replaceText,
+			replaceAll
+		})
+		this._find(findField,0)
```
@YueHui do we really need to call `this._find` here? It seems it is already called inside replace method in FindCommand. Could you please check if it is really necessary?

YueHui replied:
the purpose of call `this._find` is to update the `findField` infotext,it's Inappropriate, i fixed it

antoniolucasnobar reviewed (commented):
@YueHui I like how you created the command, very flexible. It was clever!

But I do not know if it should be only one command or two (one for find and one for replace text). It feels a little strange to have only one command. 

So, I would like to know your opinion.

YueHui committed:
optimization find info

YueHui committed:
fix parameter

YueHui commented:
if split find and replace command the replace command still needs findText and replaceText parameters,it's not much more convenient for the user,only just a name difference

antoniolucasnobar committed:
[#1] Update list of languages supported

Commit:
[#1] Fix import, bump version, add peerDependency and improve compatibility

antoniolucasnobar reviewed (changes requested):
Hi @YueHui , I finally got to test this on my environment. Sorry the delay.

I fixed the `Command` import to comply with other imports in the project. I think it will not impact your use, but, please let me know.

I noted three things using the editor interface, different from version 0.1.1:
- If I try to replace all occurrences, no message is shown to the user;
- If I try to replace a text without previously search it, it performs just a search without replace. I have to click again the replace button to actually replace the text. The same occurs if I try to replace out of editor (using `editor.execute...`)
- The console throws an error every time the code does a `this.editor.execute` when running against ckeditor5 v20; But everything works. This could be a bug in this ckeditor version and it is not really important right now. I set the compatibility to >=v21, so users of the plugin should be aware it is not recomended to use it with previous versions.

Could you please check at least the two first items, so we can fix it, before merge this pull request? 

thanks in advance.

Commit:
[#1] reduce svg icon size

antoniolucasnobar commented on src/findReplace.js:
```diff
@@ -219,147 +220,43 @@ export default class FindReplace extends Plugin {
-				size++;
-			}
-			this._resetStatus();
-			replaceField.infoText = t( 'Replaced ' ) + size + t( ' times' );
```
I think this was lost in the new version. Without it, the user get no feedback. the replace method could return total number of replacements (in the normal use, would be always 1, but in the replaceAll version, it should be variable)

antoniolucasnobar commented on src/findCommand.js:
```diff
@@ -0,0 +1,152 @@
+		const model = this.editor.model;
+		const markers = Array.from( model.markers.getMarkersGroup( SEARCH_MARKER ) );
+		const sameSearch = isSameSearch( findText, markers );
+		const currentMarker = sameSearch ? markers[ this.currentSearchIndex ] : this._find( findText, 1 );
```
maybe the problem with replace not being fired is here, now that `_find` returns an object, not only the currentMarker. I think destructure may fix this.

Commit:
[#1] Fix replace issues:

1 - if replace was fired before a search, it was only performing the search
-> this was because the _find method now returns an object. destructuring it solves the issue.

2 - After a replace all operation, no feedback was given to the user
-> we already had an variable keeping track of the number of replaces. we now return it and treat accordingly

Commit:
[#1] Add new translation for match case

- missing zh-cn translation

antoniolucasnobar commented:
Hello @YueHui 

I think I fixed the problems above. Could you please test in your environment to check if everything is ok? I do not want to make a release which may not work in your environment.

About the code, I am not so happy how I did it, but it is working. Maybe you can provide a more elegant solution. 

One last thing. I plan to support case sensitive search in the future (issue #3), so, as you provided an zh-cn translation, could you please add an translation to this feature? I added a "Match case" entry in contexts.json.

One more time, thank you for your help in this project.

YueHui committed:
add translate

YueHui commented:
sorry for reply so late, i didn't open computer at weekend,it works fine in my environment ,but when replace a text without search it, it directly replace the first item and mark the second item ,is it feel a little kind of strange?i think the origin logic is fine ,first search it ,and then the user decide to replace or not

antoniolucasnobar commented:
> sorry for reply so late, i didn't open computer at weekend,it works fine in my environment ,but when replace a text without search it, it directly replace the first item and mark the second item ,is it feel a little kind of strange?i think the origin logic is fine ,first search it ,and then the user decide to replace or not

Well, I think if the user directly clicks on the replace option, in my opinion, it should replace the first occurrence. the plugin has next and previous buttons to search (maybe we need better names for them, Find Next and Find Previous).

I made some tests in Google Docs and in LibreOffice Writer. 

Google docs search as you type, so the replace works from the first hit (this was what I had in mind when I created the plugin, but I still did not test the search as you type (I do not know if it can slow down the editor).

LibreOffice Writer behaves like you described. If I open the replace dialog (CTRL+H) and hit replace button, the first time, it performs only a search for the first item. If I hit again, the it replaces the first occurrence.
I noted a difference only if I CTRL+H with some text selected, then click the replace button, it performs the replace the first time (it is like the first search was already made).

I am going to think in this behavior better tomorow, it is a little late here in Brazil.

Commit:
[#4] Plugin Structure

Commit:
[#1] replace only when search is already made

- otherwise makes a search to show the user what is going to be replaced
- fix negative numbers in find command
- support to increment in replace

antoniolucasnobar commented:
Hi @YueHui I changed behavior. If replace is triggered before search, only search is made. same behavior as Libreoffice and accordingly to your vision.

I fixed a problem with negative numbers in find (if the negative number is bigger than the number of occurrences).

I added support to replace to accept the increment argument:

- this allows to the user in the interface to use shift+enter to replace backwards
- this also allows to use the increment parameter in findReplaceCommand to replace and choose the next item to be replaced

Maybe I should support only backward or forward search/replace (increment or decrement by 1). But this model makes it more flexible.

I will release 0.2.0; If it does not work fine in your environment, please let me know (with maybe another pull request). 

thanks for your contribution!
